### PR TITLE
[WIP] Fix cmake build system

### PR DESCRIPTION
**Pull request by Rombur**
Following up on issue #125

This PR sets the version to "2.8.0" and uses `cmakedefine` instead of `define` in `p4es_config.h.in`. I have also updated the comment in `p4es_config.h.in`. This is a single commit so it can easily be reverted if you don't like this change. There are two problems with this PR:
1. I had to set the version manually. This means that you will need to update this number before doing a release. I chose 2.8.0 but I can easily change that number.
2. As far as I can tell the current code has a problem with the vtk binary options. Currently, the macro `P4EST_VTK_BINARY` is never set and so in `p4est_vtk.c`, we always go through the ASCII path. What I don't understand is that configuring `p4est` with `-Dvtk=off` does not compile. I get the error `undefined reference to 'p4est_vtk_write_file'` in several tests. Maybe I missunderstand what the option `vtk` is supposed to do but when this option is turned off the code does not compile. With this PR  `P4EST_VTK_BINARY` is defined if `vtk=on` but the code does crash if you turn if off.

Note that currently `p4est_config.h` always report that p4est `P4EST_ENABLE_VTK_BINARY` and `P4EST_ENABLE_VTK_COMPRESSION` are defined but it is actually not true because the macros used in the code `P4EST_VTK_BINARY` and `P4EST_VTK_COMPRESSION` are never defined.

### CODE REVIEW

**Comment by cburstedde** [comment]
Thanks!

 1. With autoconf, we use git describe via build-aux/git-version-gen. Is there a way to call these from cmake?
 2. Good catch! The macros without ENABLE are deprecated and should be replaced with the ENABLE ones. Could you give this a try in p4est_vtik.c and friends and see that automake continues to work, and cmake hopefully works, too?

File: cmake/p4est_config.h.in [hunk 1]
```diff
@@ -12,42 +12,42 @@
 /* C preprocessor flags */
 #define P4EST_CPPFLAGS @P4EST_CPPFLAGS@
 
-/* Undefine if: disable the 2D library */
+/* Defined if the 2D library is enable */
```
Comment (tjhei) [inline cmake/p4est_config.h.in:15]:
```suggestion
/* Defined if the 2D library is enabled */
```
(here and below)

**Comment by Rombur** [comment]
> 1. With autoconf, we use git describe via build-aux/git-version-gen. Is there a way to call these from cmake?

I found that there was a `git.cmake` file that uses `build-aux/git-version-gen`. I could use that to populate the `p4est_config.h.in` but it reports a version of 1.1.1518 (I have no idea where that number comes from). Looking at the documentation from `build-aux/git-version-gen` [here](https://github.com/coreutils/gnulib/blob/master/build-aux/git-version-gen#L30-L38), it looks like this function only works with autoconf. It's possible that as long as you support both autoconf and cmake we could make it work when you create a tarball but it will fail once you stop using autoconf. 

> 2.   Good catch! The macros without ENABLE are deprecated and should be replaced with the ENABLE ones. Could you give 
this a try in p4est_vtik.c and friends and see that automake continues to work, and cmake hopefully works, too?

I have changed the macros. I still need to check if it works with automake.

I found the real reason of the problem when `vtk` is OFF. When the option is off, `p4est_vtk.cc` is not compiled (see [here](https://github.com/cburstedde/p4est/blob/feature-cmake/src/CMakeLists.txt#L5-L7)) What I have done is to rename the `vtk` option to `vtk_binary` and to always compile `p4est_vtk.cc`. Now the option enable/disable binary support instead of disabling compilation of `p4est_vtk.cc`

**Comment by Rombur** [comment]
Anything else you want me to do?

**Comment by cburstedde** [comment]
> Anything else you want me to do?

Thanks a lot, please let me have a look first.  :)

File: cmake/p4est_config.h.in [hunk 2]
```diff
@@ -12,42 +12,42 @@
...
-/* Undefine if: disable hybrid 2D+1D p6est library */
+/* Defined if the hybrid 2D+1D p6est library enabled */
 #cmakedefine P4EST_ENABLE_BUILD_P6EST
 
-/* Define to 1 if we are using debug build type (assertions and extra checks) */
+/* Defined if we are using debug build type (assertions and extra checks) */
 #cmakedefine P4EST_ENABLE_DEBUG @P4EST_ENABLE_DEBUG@
 
-/* Undefine if: use aligned malloc (optionally use --enable-memalign=<bytes>) */
+/* Defined if we are using aligned malloc (optionally use --enabled-memalign=<bytes>) */
```
Comment (cburstedde) [inline cmake/p4est_config.h.in:27]: Here, only here, it should be --enable-memalign.

File: src/CMakeLists.txt [hunk 3]
```diff
@@ -1,10 +1,7 @@
...
+target_sources(p4est PRIVATE p4est_base.c p4est_connectivity.c p4est.c p4est_bits.c p4est_search.c p4est_build.c p4est_algorithms.c p4est_communication.c p4est_ghost.c p4est_nodes.c p4est_points.c p4est_geometry.c p4est_iterate.c p4est_lnodes.c p4est_mesh.c p4est_balance.c p4est_io.c p4est_connrefine.c p4est_wrap.c p4est_plex.c p4est_empty.c p4est_vtk.c)
 if(P4EST_HAVE_WINSOCK2_H)
   target_link_libraries(p4est PRIVATE ${WINSOCK_LIBRARIES})
 endif()
-if(vtk)
-  target_sources(p4est PRIVATE p4est_vtk.c)
-endif()
 
 if(enable_p8est)
 target_sources(p8est PRIVATE p8est_connectivity.c p8est.c p8est_bits.c p8est_search.c p8est_build.c p8est_algorithms.c p8est_communication.c p8est_ghost.c p8est_nodes.c p8est_vtk.c p8est_points.c p8est_geometry.c p8est_iterate.c p8est_lnodes.c p8est_mesh.c p8est_tets_hexes.c p8est_balance.c p8est_io.c p8est_connrefine.c p8est_wrap.c p8est_plex.c p8est_empty.c)
```
Comment (cburstedde) [inline src/CMakeLists.txt:7]: Could you also add p8est_vtk.c (same reason as with p4est_vtk.c)?

**Comment by cburstedde** [comment]
Thanks for your work! I like this for improved consistency between autoconf and cmake.

One thing I noticed is that autoconf defines the config defines to 1 (which we do not exploit in the code).
Would it make sense to cmakedefine to 1 instead of nothing for added similarity?

And, the whole thing would be great to be applied to the sc library as well, on the prev3-develop branch.

> > 1. With autoconf, we use git describe via build-aux/git-version-gen. Is there a way to call these from cmake?
> 
> I found that there was a `git.cmake` file that uses `build-aux/git-version-gen`. I could use that to populate the `p4est_config.h.in` but it reports a version of 1.1.1518 (I have no idea where that number comes from). Looking at the documentation from `build-aux/git-version-gen` [here](https://github.com/coreutils/gnulib/blob/master/build-aux/git-version-gen#L30-L38), it looks like this function only works with autoconf. It's possible that as long as you support both autoconf and cmake we could make it work when you create a tarball but it will fail once you stop using autoconf.

Hmm this is weird. the git-version-gen works quite standalone, it just has this strange first argument that is a file that does not exist usually. For me it returns the correct version when called from the source directory (possible cwd issue?)

> > 2. Good catch! The macros without ENABLE are deprecated and should be replaced with the ENABLE ones. Could you give
> >    this a try in p4est_vtik.c and friends and see that automake continues to work, and cmake hopefully works, too?
> 
> I have changed the macros. I still need to check if it works with automake.
> 
> I found the real reason of the problem when `vtk` is OFF. When the option is off, `p4est_vtk.cc` is not compiled (see [here](https://github.com/cburstedde/p4est/blob/feature-cmake/src/CMakeLists.txt#L5-L7)) What I have done is to rename the `vtk` option to `vtk_binary` and to always compile `p4est_vtk.cc`. Now the option enable/disable binary support instead of disabling compilation of `p4est_vtk.cc`

These are really helpful changes. Thanks.

**Comment by Rombur** [comment]
> One thing I noticed is that autoconf defines the config defines to 1 (which we do not exploit in the code).

Fixed 

> Hmm this is weird. the git-version-gen works quite standalone, it just has this strange first argument that is a file that does not exist usually. For me it returns the correct version when called from the source directory (possible cwd issue?)

For some reason, the tags in my directory were not updated correctly. Pulling the tags fixed the problem.

**Comment by cburstedde** [comment]
Thanks; running our CI on it.

Can I ask you to add a file doc/author_<lastname>.txt releasing your code under the FreeBSD license?

File: cmake/p4est_config.h.in (hunk 1 above)
Comment (cburstedde) [reply cmake/p4est_config.h.in]: Thanks @tjhei for helping out!

**Comment by cburstedde** [comment]
Thanks!

**Comment by cburstedde** [comment]
Sorry for the close, unintentional.

**Pull request merged**